content/courses/native-onchain-development/program-security.md

```diff
@@ -1,119 +1,131 @@
+there may be unintentional behaviors that could be caught by thinking like an
+attacker. In this lesson, we'll explore these issues and how to address them.
 
 ### Error handling
 
 Before we dive into some of the common security pitfalls and how to avoid them,
-it's important to know how to use errors in your program. While your code can
-handle some issues gracefully, other issues will require that your program stop
-execution and return a program error.
+it's important to know how to use errors in your program. Security issues in a
+Solana program often requires terminating the execution with a meaningful error.
+Not all errors are catastrophic, but some should result in stopping the program
+and returning an appropriate error code to prevent further processing.
 
-#### How to create errors
+#### Creating Custom Errors
 
-While the `solana_program` crate provides a `ProgramError` enum with a list of
-generic errors we can use, it will often be useful to create your own. Your
-custom errors will be able to provide more context and detail while you're
-debugging your code.
+Solana's
+[`solana_program`](https://docs.rs/solana-program/latest/solana_program/) crate
+provides a generic
+[`ProgramError`](https://docs.rs/solana-program/latest/solana_program/program_error/enum.ProgramError.html)
+enum for error handling. However, custom errors allow you to provide more
+detailed, context-specific information that helps during debugging and testing.
 
 We can define our own errors by creating an enum type listing the errors we want
 to use. For example, the `NoteError` contains variants `Forbidden` and
 `InvalidLength`. The enum is made into a Rust `Error` type by using the `derive`
-attribute macro to implement the `Error` trait from the `thiserror` library.
-Each error type also has its own `#[error("...")]` notation. This lets you
-provide an error message for each particular error type.
+attribute macro to implement the `Error` trait from the
+[`thiserror`](https://docs.rs/thiserror/latest/thiserror/) library. Each error
+type also has its own `#[error("...")]` notation. This lets you provide an error
+message for each particular error type.
+
+Here's an example of how you can define custom errors in your program:
 
 ```rust
-use solana_program::{program_error::ProgramError};
+use solana_program::program_error::ProgramError;
 use thiserror::Error;
 
-#[derive(Error)]
+#[derive(Error, Debug)]
 pub enum NoteError {
-    #[error("Wrong note owner")]
+    #[error("Unauthorized access - You don't own this note.")]
     Forbidden,
 
-    #[error("Text is too long")]
+    #[error("Invalid note length - The text exceeds the allowed limit.")]
     InvalidLength,
 }
 ```
 
-#### How to return errors
+<Callout>
+
+In this example, we create custom errors for unauthorized access and invalid
+data input (such as note length). Defining custom errors gives us greater
+flexibility when debugging or explaining what went wrong during execution.
+</Callout>
```

Comment:
```suggestion
```

Yes good errors help us debug, that should be obvious. I'm not sure this warrants a callout.

Reply:
Removed callout as it's not that important.